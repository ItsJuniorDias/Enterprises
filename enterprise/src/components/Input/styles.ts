import styled, { css } from 'styled-components/native';
import FeatherIcon from 'react-native-vector-icons/Feather';

interface ContainerProps {
  isFocused: boolean;
  isErrored: boolean;
  isValue: string;
}

export const Container = styled.View<ContainerProps>`
  width: 100%;
  height: 60px;
  padding: 0 16px;
  background: #f3f3f3;
  border-radius: 10px;
  margin-bottom: 16px;

  border-width: 2px;
  border-color: #f3f3f3;
  flex-direction: row;
  align-items: center;

  ${(props) =>
    props.isErrored &&
    css`
      border-color: #c53030;
    `}

  ${(props) =>
    props.isValue &&
    css`
      border-color: #f3f3f3;
    `}

  ${(props) =>
    props.isFocused &&
    css`
      border-color: #2d4379;
    `}
`;

export const TextInput = styled.TextInput`
  flex: 1;
  color: #9d9d9d;
  font-size: 16px;
  font-family: 'Poppins-Regular';
`;

export const Icon = styled(FeatherIcon)`
  margin-right: 16px;
`;
