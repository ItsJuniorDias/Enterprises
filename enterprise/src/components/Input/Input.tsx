import React, {
  useState,
  useEffect,
  useRef,
  useCallback,
  useImperativeHandle,
  forwardRef,
} from 'react';
import { TextInputProps } from 'react-native';
import { useField } from '@unform/core';

import { Container, TextInput, Icon } from './styles';
interface InputProps extends TextInputProps {
  name: string;
  icon: string;
}

interface InputValueReferences {
  value: string;
}

interface InputRef {
  focus(): void;
}

const Input: React.RefForwardingComponent<InputRef, InputProps> = (
  { testID = 'inputId', name, icon, ...rest },
  ref
) => {
  const inputElementRef = useRef<any>(null);
  const { registerField, defaultValue = '', fieldName, error } = useField(name);

  const inputValueRef = useRef<InputValueReferences>({ value: defaultValue });

  const [isFocused, setIsFocused] = useState(false);
  const [isFilled, setIsFilled] = useState(false);

  const handleInputFocus = useCallback(() => {
    setIsFocused(true);
  }, []);

  const handleInputBlur = useCallback(() => {
    setIsFocused(false);

    setIsFilled(!!inputValueRef.current.value);
  }, []);

  useImperativeHandle(ref, () => ({
    focus() {
      inputElementRef.current.focus();
    },
  }));

  useEffect(() => {
    registerField<string>({
      name: fieldName,
      ref: inputValueRef.current,
      path: 'value',
      setValue(ref: any, value) {
        inputValueRef.current.value = value;

        inputElementRef.current.setNativeProps({ text: value });
      },
      clearValue() {
        inputValueRef.current.value = '';
        inputElementRef.current.clear();
      },
    });
  }, [fieldName, registerField]);

  return (
    <Container
      testID="container_inputId"
      isFocused={isFocused}
      isErrored={!!error}
      isValue={inputValueRef.current.value}
    >
      <Icon
        name={icon}
        size={20}
        color={isFocused || isFilled ? '#2d4379' : '#9d9d9d'}
      />

      <TextInput
        testID={testID}
        ref={inputElementRef}
        keyboardAppearance="light"
        placeholderTextColor="#9d9d9d"
        defaultValue={defaultValue}
        onFocus={handleInputFocus}
        onBlur={handleInputBlur}
        onChangeText={(value) => {
          inputValueRef.current.value = value;
        }}
        {...rest}
      />
    </Container>
  );
};

export default forwardRef(Input);
