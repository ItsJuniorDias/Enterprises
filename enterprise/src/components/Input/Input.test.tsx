import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';

import Input from '../Input/Input';

jest.mock('react-native-vector-icons/Feather', () => 'Icon');
jest.mock('@unform/core', () => {
  return {
    useField: jest.fn(() => ({
      registerField: jest.fn(),
    })),
  };
});

describe('Behavior Input', () => {
  const screenRender = (
    <Input
      autoCorrect={false}
      autoCapitalize="none"
      keyboardType="email-address"
      name="email"
      icon="mail"
      placeholder="E-mail"
      returnKeyType="next"
    />
  );

  it('render snapshot', () => {
    const result = render(screenRender).toJSON();

    expect(result).toMatchSnapshot();
  });

  it('should call the onFocus', () => {
    const { getByTestId } = render(screenRender);

    const input = getByTestId('inputId');
    const container = getByTestId('container_inputId');

    fireEvent(input, 'focus');

    expect(container.props.style).toMatchObject([
      {
        borderColor: '#2d4379',
      },
    ]);
  });

  it('should call the onChange function', () => {
    const { getByTestId } = render(screenRender);

    const input = getByTestId('inputId');

    fireEvent.changeText(input, 'juniordias@gmail.com');

    expect(input.props.onChangeText()).toBeUndefined();
  });

  it('should call the onBlur', () => {
    const { getByTestId } = render(screenRender);

    const input = getByTestId('inputId');

    fireEvent(input, 'blur');

    expect(input.props.onBlur()).toBeUndefined();
  });
});
