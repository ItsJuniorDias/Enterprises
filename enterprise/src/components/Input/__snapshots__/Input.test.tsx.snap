// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Behavior Input render snapshot 1`] = `
<View
  isErrored={false}
  isFocused={false}
  isValue=""
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#f3f3f3",
        "borderColor": "#f3f3f3",
        "borderRadius": 10,
        "borderWidth": 2,
        "flexDirection": "row",
        "height": 60,
        "marginBottom": 8,
        "paddingBottom": 0,
        "paddingLeft": 16,
        "paddingRight": 16,
        "paddingTop": 0,
        "width": "100%",
      },
    ]
  }
  testID="container_inputId"
>
  <Icon
    color="#9d9d9d"
    name="mail"
    size={20}
    style={
      Array [
        Object {
          "marginRight": 16,
        },
      ]
    }
  />
  <TextInput
    allowFontScaling={true}
    autoCapitalize="none"
    autoCorrect={false}
    defaultValue=""
    keyboardAppearance="light"
    keyboardType="email-address"
    onBlur={[Function]}
    onChangeText={[Function]}
    onFocus={[Function]}
    placeholder="E-mail"
    placeholderTextColor="#9d9d9d"
    rejectResponderTermination={true}
    returnKeyType="next"
    style={
      Array [
        Object {
          "color": "#9d9d9d",
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "fontFamily": "Poppins-Regular",
          "fontSize": 16,
        },
      ]
    }
    testID="inputId"
    underlineColorAndroid="transparent"
  />
</View>
`;
